### PR TITLE
Add voting data structures

### PR DESCRIPTION
**Pull request by MichaelKim20**
This is the code that defines the data related to Votera's proposal and voting.
This includes serialization and deserialization codes for data as binaries.

Related to #105

### CODE REVIEW

**Comment by MichaelKim20** [comment]
I rebased it

**Comment by MichaelKim20** [comment]
The data structures included will be used in common with Agora and Stoa.
So I implemented serialization and deserialization similar to Agora.

File: src/modules/vote/BallotData.ts [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+    /**
+     * The id of the proposal
+     */
+    public proposal_id: string;
+
+    /**
+     * Encrypted voting information
+     */
+    public ballot: Buffer;
```
Comment (MichaelKim20) [inline src/modules/vote/BallotData.ts:130]: The data structures included will be used in common with Agora and Stoa.
So I implemented serialization and deserialization similar to Agora.

File: src/modules/vote/BallotData.ts [hunk 2]
```diff
@@ -144,6 +144,10 @@ export class BallotData
      */
     public signature: Signature;
 
+    public static YES: string   = "Yes  ";
+    public static NO: string    = "No   ";
+    public static BLANK: string = "Blank";
```
Comment (MichaelKim20) [inline src/modules/vote/BallotData.ts:149]: The length of the message is equal to the length of the message to ensure the same length after encryption.

File: src/modules/vote/Encrypt.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        return SodiumHelper.sodium.crypto_generichash_final(state, key_size);
+    }
+
+    /**
+     * Encrypts a message using a secret key and public nonce
+     * @param message       The message to encrypt
+     * @param key           The secret key
+     * @param proposal_id   The public nonce
+     */
+    static encrypt (message: Buffer, key: Buffer, proposal_id: string)
```
Comment (MichaelKim20) [inline src/modules/vote/Encrypt.ts:43]: I'm thinking about what to make of a public nonce.

File: src/modules/vote/ProposalData.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    /**
+     * The type of the proposal, 0: System, 1: Funding
+     */
+    public proposal_type: ProposalType;
+
+    /**
+     * The id of the proposal
+     */
+    public proposal_id: string;
```
Comment (MichaelKim20) [inline src/modules/vote/ProposalData.ts:46]: The title of the proposal must be added.

**Comment by MichaelKim20** [comment]
The ability to calculate fees for a single vote was added to the final commitment.

File: src/modules/vote/Encrypt.ts (hunk 3 above)
Comment (MichaelKim20) [reply src/modules/vote/Encrypt.ts]: I changed to using random bytes in the nonce.

**Comment by MichaelKim20** [comment]
@Geod24 Some of the content in this PR should be implemented in Agora.
If you have any opinion about this PR, please leave it.

File: src/modules/utils/VarInt.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+ * when serializing and deserializing them in binary.
+ */
+export class VarInt
+{
+    /**
+     * Serialize the number
+     * @param value The number to serialize
+     * @param buffer The byte buffer
+     */
+    public static serializeFromNumber (value: number, buffer: SmartBuffer)
```
Comment (Geod24) [inline src/modules/utils/VarInt.ts:33]: Tiny bit of style advice:
If all your method start with the same prefix, you don't need the prefix. In this case:
- `serializeFromNumber` => `fromNumber`;
- `serializeFromJSBI` => fromJSBI`;
- `deserializeToNumber` => `toNumber`;
- `deserializeToJSBI` => `toJSBI`;

File: src/modules/vote/BallotData.ts [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        return new VoterCard(validator_address, address, expires, signature);
+    }
+}
+
+/**
+ * voting data stored in transaction payloads
+ */
+export class BallotData
+{
+    public static HEADER = "BALLOT  "
```
Comment (Geod24) [inline src/modules/vote/BallotData.ts:120]: What is this ?

File: src/modules/vote/BallotData.ts (hunk 2 above)
Comment (Geod24) [reply src/modules/vote/BallotData.ts:151]: We're going to use an integer, not a string. I'll amend the specs.

File: src/modules/vote/ProposalData.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+    /**
+     * The block height of voting start
+     */
+    public vote_start_height: JSBI;
+
+    /**
+     * The block height of voting end
+     */
+    public vote_end_height: JSBI;
```
Comment (Geod24) [inline src/modules/vote/ProposalData.ts:61]: Where does all this come from ?

File: src/modules/vote/BallotData.ts (hunk 6 above)
Comment (MichaelKim20) [reply src/modules/vote/BallotData.ts:120]: Payloads can store many kinds of data. I decided that I needed a header to distinguish them.
For that reason, I added headers for each of the three types of data.

File: src/modules/vote/ProposalData.ts (hunk 7 above)
Comment (MichaelKim20) [reply src/modules/vote/ProposalData.ts:61]: This data is passed from the Votera app to Agora.
It is also delivered to Agora after the pre-examination of the proposal is completed.
The transaction sent at that time has the data below.
1. Transaction outputs that pay for voting to all validators.
2. Payload where the above data is stored

Comment (MichaelKim20) [reply src/modules/vote/ProposalData.ts:61]: @Geod24 I just added another commit.
I added a test code on how to verify the proposed data and the transaction and payload related to the ballot at the Voter Server.
Please check.

File: src/modules/vote/BallotData.ts [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        this.card.serialize(buffer);
+        VarInt.serializeFromNumber(this.sequence, buffer);
+        buffer.writeBuffer(this.signature.data);
+    }
+
+    /**
+     * Deserialize from binary data.
+     * @param buffer The buffer to be deserialized
+     */
+    public static deserialize (buffer: SmartBuffer): BallotData
```
Comment (MichaelKim20) [inline src/modules/vote/BallotData.ts:220]: A code must be added to check that the size of the buffer entered is the expected size.

File: src/modules/vote/ProposalData.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        buffer.writeBuffer(this.tx_hash_proposal_fee.data);
+        buffer.writeBuffer(this.proposer_address.data);
+        buffer.writeBuffer(this.proposal_fee_address.data);
+    }
+
+    /**
+     * Deserialize as binary data.
+     * @param buffer The buffer to be deserialized
+     */
+    public static deserialize (buffer: SmartBuffer): ProposalData
```
Comment (MichaelKim20) [inline src/modules/vote/ProposalData.ts:170]: A code must be added to check that the size of the buffer entered is the expected size.

File: src/modules/vote/ProposalFeeData.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        temp = Buffer.from(this.proposal_id);
+        VarInt.serializeFromNumber(temp.length, buffer);
+        buffer.writeBuffer(temp);
+    }
+
+    /**
+     * Deserialize as binary data.
+     * @param buffer The buffer to be deserialized
+     */
+    public static deserialize (buffer: SmartBuffer): ProposalFeeData
```
Comment (MichaelKim20) [inline src/modules/vote/ProposalFeeData.ts:60]: A code must be added to check that the size of the buffer entered is the expected size.

**Comment by MichaelKim20** [comment]
@Geod24 
I revised everything you suggested. 
Is there anything else that needs to be modified?

**Comment by MichaelKim20** [comment]
I will merge this first, and I will do additional PR if any modifications occur later.

**Pull request merged**